--- feeds/node/node/Makefile
+++ feeds/node/node/Makefile
@@ -180,8 +180,10 @@
 endef

 define Package/node-npm/install
-	$(INSTALL_DIR) $(1)/usr/lib/node_modules
-	$(CP) $(PKG_INSTALL_DIR)/usr/lib/node_modules/* $(1)/usr/lib/node_modules/
+	$(INSTALL_DIR) $(1)/usr/lib/node_modules/npm/bin
+	$(CP) $(PKG_INSTALL_DIR)/usr/lib/node_modules/npm/bin/{npm,npx,*.js} $(1)/usr/lib/node_modules/npm/bin/
+	$(CP) $(PKG_INSTALL_DIR)/usr/lib/node_modules/npm/{lib,scripts,node_modules} $(1)/usr/lib/node_modules/npm/
+	$(CP) $(PKG_INSTALL_DIR)/usr/lib/node_modules/npm/{LICENSE,AUTHORS,package.json} $(1)/usr/lib/node_modules/npm/
 	$(INSTALL_DIR) $(1)/usr/bin
 	$(LN) ../lib/node_modules/npm/bin/npm-cli.js $(1)/usr/bin/npm
 	$(LN) ../lib/node_modules/npm/bin/npx-cli.js $(1)/usr/bin/npx
