--- package/feeds/node/node-zigbee2mqtt/Makefile
+++ package/feeds/node/node-zigbee2mqtt/Makefile
@@ -6,12 +6,12 @@

 PKG_NPM_NAME:=zigbee2mqtt
 PKG_NAME:=node-$(PKG_NPM_NAME)
-PKG_VERSION:=1.18.1
+PKG_VERSION:=1.18.2
 PKG_RELEASE:=1

 PKG_SOURCE:=$(PKG_NPM_NAME)-$(PKG_VERSION).tgz
 PKG_SOURCE_URL:=https://registry.npmjs.org/$(PKG_NPM_NAME)/-/
-PKG_HASH:=6a2bf71a3d6e73d8da167e4f87af341d59f4cf40e94d2eefefc3c79d708a52f4
+PKG_HASH:=d30015b4ba8e02f72464a676e770fb58216dbc402a15b8beaad633b891c8faa4

 PKG_MAINTAINER:=Hirokazu MORIKAWA <morikw2@gmail.com>
 PKG_LICENSE:=GPL-3.0
