--- package/feeds/node/node-zigbee2mqtt/Makefile
+++ package/feeds/node/node-zigbee2mqtt/Makefile
@@ -6,12 +6,12 @@

 PKG_NPM_NAME:=zigbee2mqtt
 PKG_NAME:=node-$(PKG_NPM_NAME)
-PKG_VERSION:=1.18.2
+PKG_VERSION:=1.18.3
 PKG_RELEASE:=1

 PKG_SOURCE:=$(PKG_NPM_NAME)-$(PKG_VERSION).tgz
 PKG_SOURCE_URL:=https://registry.npmjs.org/$(PKG_NPM_NAME)/-/
-PKG_HASH:=d30015b4ba8e02f72464a676e770fb58216dbc402a15b8beaad633b891c8faa4
+PKG_HASH:=d7f03114f7226417cd47e56f429f9f6c0ab12bb103ac9e497a923ad3cef2f3e5

 PKG_MAINTAINER:=Hirokazu MORIKAWA <morikw2@gmail.com>
 PKG_LICENSE:=GPL-3.0
